tools/oesign/main.c

```diff
@@ -468,11 +474,16 @@ static const char _usage_gen[] =
 static const char _usage_sign[] =
     "Usage: %s sign {--enclave-image | -e} ENCLAVE_IMAGE "
     "{--config-file | -c} CONFIG_FILE {--key-file | -k} KEY_FILE\n"
+    "{--engine ENGINE-NAME -load-path ENGINE-LOAD-PATH-key KEY-ID }\n"
```

Comment:
`--load-path` and `--key` for consistency with parameter naming. The parser also appears to support short versions of the parameter names (`-e`, `-l` and `-i` respectively) but that's not reflected here. 

I'm curious why we decided to accept the load path on the command line, it seems like the approach most users would be familiar with would be as done through the openssl commands (e.g. `openssl dgst -engine "myengine"` where `-inkey` is overloaded to accept a key ID).

Reply:
Also consider if it might be cleaner if KEY_FILE could refer to a key configuration file that encapsulated the engine and key ID parameters. It avoids additional logic to manage the mutual exclusion checking and that multiple parameters must be specified with `--engine`. We already have a parser for name value pairs via CONFIG_FILE after all.

Reply:
Loadpath is a standard parameter for connecting to dynamic engines and required unless the engine is configured into openssl.conf.  The documentation of the subject shows the process using -pre argments. 

The key-id isn't a key, its a string with meaning only to the engine.

Overloading key-file to mean either a pem or a completely different kind of file seems to complicate the semantics.  

Wouldn't the idea of a configuration file to indicate an engine/loadpath/keyid triple would obfuscate what was going on on the command line, as you would not be able to see the engine parameters in the build. In fact you would not see that you were using an engine at all unless you happened to look at the (probably cmake-generated) configuration file.

Reply:
@CodeMonkeyLeet  how would you deduce that the key-file parameter was in fact a key, or a config file containing the requisite triple of engineid/keyid/engineloadpath?  Would it be restricted to .pem files, with anything else treated as an engine config file?

I presume then that all oesign using cmake file would have to be modified to create the config file.

Wouldn't that impose a limitation on oesign that isn't there currently?  I do understand that the code as written prefers pem files, but they don't have to be named *.pem

Reply:
> Also consider if it might be cleaner if KEY_FILE could refer to a key configuration file that encapsulated the engine and key ID parameters. It avoids additional logic to manage the mutual exclusion checking and that multiple parameters must be specified with --engine. We already have a parser for name value pairs via CONFIG_FILE after all.

I think this makes embedding the tool into another script cumbersome. With engine parameters on the command line, the wrapper script can simply invoke **oesign** with the correct parameters. But with the engine configuration file, the wrapper script now needs to generate a file, manage the location of the file, invoke **oesign** and then delete the file. I really think it would be more convenient to achieve the same effect with command-line options if possible.

Reply:
fixed help text as requested.